packages/react-charting/src/utilities/utilities.ts

```diff
@@ -1109,22 +1110,96 @@ export function wrapTextInsideDonut(selectorClass: string, maxWidth: number) {
+
+    // Determine the ellipsis length for word truncation
+    const ellipsis = '...';
+    tspan.text(ellipsis);
+    const ellipsisLength = tspan.node()!.getComputedTextLength();
+    tspan.text(null);
     while ((word = words.pop()!)) {
       line.push(word);
       tspan.text(line.join(' ') + ' ');
-      if (tspan.node()!.getComputedTextLength() > maxWidth && line.length > 1) {
+      // Determine if wrapping is required
+      // If yes, append a new line only if it does not exceed the max height
+      // here, max width = max height, so maxWidth is reused to check max height
+      if (
+        tspan.node()!.getComputedTextLength() > maxWidth &&
+        line.length > 1 &&
+        tspan.node()!.getBoundingClientRect().y < maxWidth
+      ) {
         line.pop();
         tspan.text(line.join(' ') + ' ');
-        line = [word];
-        tspan = text
-          .append('tspan')
-          .attr('id', `WordBreakId-${idx}-${lineNumber}`)
-          .attr('x', 0)
-          .attr('y', y)
-          .attr('dy', lineNumber++ * lineHeight + 'em')
-          .text(word);
+        // append a line only if the line is within the max height
+        // else truncate horizontally and break
+        if (tspan.node()!.getBoundingClientRect().y + lineLength < maxWidth) {
+          line = [word];
+          tspan = text
+            .append('tspan')
+            .attr('id', `WordBreakId-${idx}-${lineNumber}`)
+            .attr('x', 0)
+            .attr('y', y)
+            .attr('dy', lineNumber++ * lineHeight + 'em')
+            .text(word);
+        } else {
+          truncateHorizontally(tspan, line, maxWidth, ellipsisLength);
+          break;
+        }
+      }
+      // Determine if truncation is required vertically
+      // If yes, remove the last line, append ellipsis
+      // to the last word of the previous line (truncate horizontally) and break
+      if (tspan.node()!.getBoundingClientRect().y > maxWidth && line.length >= 1) {
+        line.pop();
+        tspan.text(line.join(' '));
+        truncateHorizontally(tspan, line, maxWidth, ellipsisLength);
+        break;
+      }
+      // Determine if truncation is required horizontally
+      if (tspan.node()!.getComputedTextLength() > maxWidth && line.length >= 1) {
+        truncateHorizontally(tspan, line, maxWidth, ellipsisLength);
+        break;
       }
     }
     idx += 1;
   });
 }
+
+function truncateHorizontally(
+  tspan: Selection<SVGTSpanElement, unknown, null, undefined>,
+  line: string[],
+  maxWidth: number,
+  ellipsisLength: number,
+) {
+  const ellipsis = '...';
+  let lastWord = '';
+  // atleast 3 chars to remove ('...')
+  // may require more removal based on the locale alphabets
+  const truncateCount = 4;
+  // Truncate till the text fits in the maximum allowable width
+  while (tspan.node()!.getComputedTextLength() > maxWidth - ellipsisLength) {
+    lastWord = line.pop()!;
+    if (lastWord !== undefined && lastWord.length >= truncateCount) {
```

Comment:
shouldn't break happen only if last word is undefined. Is second part of the expression also relevant.

Reply:
yes this is relevant because if the (tspan.node()!.getComputedTextLength() > maxWidth - ellipsisLength) and the lastword.length is < truncateCount, popping just the last word is sufficient after which it will append the ellipsis